File: packages/server/src/store.ts
/**
 * @mbsks/openrtc-server — the `Store` contract.
 *
 * The server core is a set of pure functions (`core.ts`) that take a
 * `Store` as their first argument. A `Store` is a minimal KV + query
 * surface — ~10 methods — that any database can implement:
 *
 *  - rooms:                 one row per room (JSON document)
 *  - participants:          one row per (room, participant)
 *  - signals:               append-only per-room log of protocol envelopes
 *  - recordings:            one row per recording session (metadata only)
 *
 * Implementations: `InMemoryStore` ships from the default entry; the
 * SQL-backed ones live behind subpath exports so the core stays
 * driver-free — `@mbsks/openrtc-server/stores/sqlite` (better-sqlite3),
 * `/stores/postgres` (pg), `/stores/mysql` (mysql2). See
 * `integrations/DATABASES.md` for how to implement one for any other
 * database (MongoDB, DynamoDB, Redis, Firestore, ...).
 *
 * Contract notes:
 *  - All methods are async (even in-memory) so implementations stay uniform.
 *  - `putSignal` returns the stored signal: the Store assigns the
 *    per-room monotonic `seq` atomically (identity column / MAX+1 / counter).
 *  - JSON documents round-trip verbatim (the Store must not reorder or
 *    drop fields).
 */

import type { Envelope } from '@mbsks/openrtc-protocol';
import type { BanEntry, LobbyEntry, Participant, Poll, RecordingSession, Room, StoredSignal } from './types.ts';

/** A signal waiting to be persisted (seq is assigned by the Store). */
export interface SignalInput {
  roomId: string;
  envelope: Envelope;
  /** Epoch ms when the server accepted the signal. */
  receivedAt: number;
}

/** Minimal KV + query surface implemented by every backing database. */
export interface Store {
  // ---- rooms -------------------------------------------------------------
  getRoom(roomId: string): Promise<Room | null>;
  putRoom(room: Room): Promise<void>;
  /** Remove a room and its participants (used by closeRoom + tests). */
  deleteRoom(roomId: string): Promise<void>;

  // ---- participants ------------------------------------------------------
  getParticipant(roomId: string, participantId: string): Promise<Participant | null>;
  putParticipant(participant: Participant): Promise<void>;
  deleteParticipant(roomId: string, participantId: string): Promise<void>;
  listParticipants(roomId: string): Promise<Participant[]>;

  // ---- signals (append-only per-room log) --------------------------------
  putSignal(signal: SignalInput): Promise<StoredSignal>;
  /** Signals with `seq > since`, ordered ascending by seq. */
  listSignals(roomId: string, since: number): Promise<StoredSignal[]>;

  // ---- recordings --------------------------------------------------------
  putRecording(recording: RecordingSession): Promise<void>;
  listRecordings(roomId: string): Promise<RecordingSession[]>;
  getRecording(sessionId: string): Promise<RecordingSession | null>;
  deleteRecording?(sessionId: string): Promise<void>;
  listAllRecordings?(): Promise<RecordingSession[]>;

  // ---- bans (per-room) ---------------------------------------------------
  listBans?(roomId: string): Promise<BanEntry[]>;
  getBan?(roomId: string, participantId: string): Promise<BanEntry | null>;
  putBan?(roomId: string, entry: BanEntry): Promise<void>;
  deleteBan?(roomId: string, participantId: string): Promise<void>;

  // ---- lobby (waiting queue) ---------------------------------------------
  listLobby?(roomId: string): Promise<LobbyEntry[]>;
  putLobby?(roomId: string, participantId: string, enqueuedAt: number): Promise<void>;
  deleteLobby?(roomId: string, participantId: string): Promise<boolean>;

  // ---- hand queue (ordered) ----------------------------------------------
  listHandQueue?(roomId: string): Promise<string[]>;
  addHand?(roomId: string, participantId: string): Promise<void>;
  removeHand?(roomId: string, participantId: string): Promise<void>;

  // ---- polls -------------------------------------------------------------
  listPolls?(roomId: string): Promise<Poll[]>;
  getPoll?(roomId: string, pollId: string): Promise<Poll | null>;
  putPoll?(roomId: string, poll: Poll): Promise<void>;
  votePoll?(roomId: string, pollId: string, participantId: string, option: string): Promise<boolean>;
}

/** Shape a store implementation must satisfy — re-exported for convenience. */
export type StoreLike = Store;

File: server/rust/crates/vidcall-server/src/store.rs
//! The [`Store`] contract — mirror of `packages/server/src/store.ts`.
//!
//! The server core is a set of pure functions (`core.rs`) that take a
//! `Store` as their first argument. A `Store` is a minimal KV + query
//! surface — ~10 methods — that any database can implement:
//!
//!  - rooms:            one row per room (JSON document)
//!  - participants:     one row per (room, participant)
//!  - signals:          append-only per-room log of protocol envelopes
//!  - recordings:       one row per recording session (metadata only)
//!
//! Implementations ship in `stores/`: `InMemoryStore` (tests/dev),
//! `SqliteStore` (rusqlite, bundled), `PostgresStore` (tokio-postgres),
//! `ConvexStore` (Convex HTTP actions), `SupabaseStore` (PostgREST REST CRUD)
//! and the generic `HttpJsonStore` pattern (Firebase/Appwrite/any BaaS).
//! Every implementation passes the SHARED store test suite
//! (`crate::shared_tests::run_store_test_suite`), mirroring the client-side
//! shared-adapter-suite pattern.
//!
//! Contract notes:
//!  - All methods are async (even in-memory) so implementations stay uniform.
//!  - `put_signal` returns the stored signal: the Store assigns the per-room
//!    monotonic `seq` atomically (identity column / MAX+1 / counter).
//!  - JSON documents round-trip verbatim (the Store must not reorder or drop
//!    fields).

use async_trait::async_trait;
use futures_util::Stream;

use crate::protocol::Envelope;
use crate::types::{Participant, RecordingSession, Room, StoredSignal};

/// A signal waiting to be persisted (seq is assigned by the Store).
#[derive(Debug, Clone)]
pub struct SignalInput {
    pub room_id: String,
    pub envelope: Envelope,
    /// Epoch ms when the server accepted the signal.
    pub received_at: i64,
}

/// A stream of stored signals (the change feed for a room). Boxed so
/// non-`Unpin` stream implementations (async pollers) can be returned.
pub type SignalStream = Box<dyn Stream<Item = StoredSignal> + Send>;

/// Minimal KV + query surface implemented by every backing database.
///
/// Object-safe (via `async_trait`) so a sidecar can hold
/// `Arc<dyn Store>` while library users can also keep concrete stores.
#[async_trait]
pub trait Store: Send + Sync + 'static {
    // ---- rooms -----------------------------------------------------------

    async fn get_room(&self, room_id: &str) -> crate::error::Result<Option<Room>>;

    async fn put_room(&self, room: &Room) -> crate::error::Result<()>;

    /// Remove a room and its participants (used by `close_room` + tests).
    /// Optional: the default implementation is a no-op, matching the TS
    /// `deleteRoom?` optional method.
    async fn delete_room(&self, _room_id: &str) -> crate::error::Result<()> {
        Ok(())
    }

    // ---- participants ----------------------------------------------------

    async fn get_participant(
        &self,
        room_id: &str,
        participant_id: &str,
    ) -> crate::error::Result<Option<Participant>>;

    async fn put_participant(&self, participant: &Participant) -> crate::error::Result<()>;

    async fn delete_participant(
        &self,
        room_id: &str,
        participant_id: &str,
    ) -> crate::error::Result<()>;

    async fn list_participants(&self, room_id: &str) -> crate::error::Result<Vec<Participant>>;

    // ---- signals (append-only per-room log) ------------------------------

    async fn put_signal(&self, signal: SignalInput) -> crate::error::Result<StoredSignal>;

    /// Signals with `seq > since`, ordered ascending by seq.
    async fn list_signals(&self, room_id: &str, since: i64)
        -> crate::error::Result<Vec<StoredSignal>>;

    // ---- recordings ------------------------------------------------------

    async fn put_recording(&self, recording: &RecordingSession) -> crate::error::Result<()>;

    async fn list_recordings(&self, room_id: &str) -> crate::error::Result<Vec<RecordingSession>>;

    async fn get_recording(&self, session_id: &str)
        -> crate::error::Result<Option<RecordingSession>>;

    // ---- change feed (WS relay / cross-node fan-out) ---------------------

    /// Subscribe to the room's signal change feed.
    ///
    /// Returns `None` when the store has no feed (custom stores may leave the
    /// default). In-memory/SQLite stores return an in-process broadcast;
    /// remote stores (Postgres/Convex/Supabase/HttpJson) return a polling
    /// stream over `list_signals`. The sidecar's WebSocket relay drives its
    /// in-process hub directly; `subscribe` is the cross-process mechanism —
    /// see `DATABASES.md`.
    fn subscribe(&self, _room_id: &str) -> Option<SignalStream> {
        None
    }
}
